napari/utils/colormaps/colormap_utils.py

```diff
@@ -472,6 +474,92 @@ def label_colormap(
     )
 
 
+@lru_cache
+def _primes(upto=2**16):
+    """Generate primes up to a given number.
+
+    Parameters
+    ----------
+    upto : int
+        The upper limit of the primes to generate.
+
+    Returns
+    -------
+    primes : np.ndarray
+        The primes up to the upper limit.
+    """
+    primes = np.arange(3, upto + 1, 2)
+    isprime = np.ones((upto - 1) // 2, dtype=bool)
+    max_factor = int(np.sqrt(upto))
+    for factor in primes[: max_factor // 2]:
+        if isprime[(factor - 2) // 2]:
+            isprime[(factor * 3 - 2) // 2 : None : factor] = 0
+    return np.concatenate(([2], primes[isprime]))
+
+
+def shuffle_and_extend_colormap(
+    colormap: LabelColormap, seed: int, min_random_choices: int = 5
+) -> LabelColormap:
+    """Shuffle the colormap colors and extend it to more colors.
+
+    The new number of colors will be a prime number that fits into the same
+    dtype as the current number of colors in the colormap.
+
+    Parameters
+    ----------
+    colormap : napari.utils.LabelColormap
+        Colormap to shuffle and extend.
+    seed : int
+        Seed for the random number generator.
+    min_random_choices : int
+        Minimum number of new table sizes to choose from. When choosing table
+        sizes, every choice gives a 1/size chance of two labels being mapped
+        to the same color. Since we try to stay within the same dtype as the
+        original colormap, if the number of original colors is close to the
+        maximum value for the dtype, there will not be enough prime numbers
+        up to the dtype max to ensure that two labels can always be
+        distinguished after one or few shuffles. In that case, we discard
+        some colors and choose the `min_random_choices` largest primes to fit
+        within the dtype.
+
+    Returns
+    -------
+    colormap : napari.utils.LabelColormap
+        Shuffled and extended colormap.
+    """
+    rng = np.random.default_rng(seed)
+    n_colors_prev = len(colormap.colors)
+    dtype = minimum_dtype_for_labels(n_colors_prev)
+    indices = np.arange(n_colors_prev)
+    rng.shuffle(indices)
+    shuffled_colors = colormap.colors[indices]
+
+    primes = _primes(np.iinfo(dtype).max)
+    valid_primes = primes[primes > n_colors_prev]
+    if len(valid_primes) < min_random_choices:
+        valid_primes = primes[-min_random_choices:]
+    n_colors = rng.choice(valid_primes)
+
+    extended_colors = np.concatenate(
+        (
+            shuffled_colors,
+            shuffled_colors[
+                rng.choice(indices, size=n_colors - n_colors_prev)
+            ],
+        ),
+        axis=0,
+    )
+
+    new_colormap = LabelColormap(
+        name=colormap.name,
+        colors=extended_colors,
+        controls=np.linspace(0, 1, len(extended_colors) + 1),
+        interpolation='zero',
+        background_value=colormap.background_value,
+    )
+    return new_colormap
```

Comment:
```suggestion
    rng = np.random.default_rng(seed)
    n_colors_prev = len(colormap.colors)
    dtype = minimum_dtype_for_labels(n_colors_prev)
    cached_colors = colormap._get_mapping_from_cache(dtype)

    primes = _primes(np.iinfo(dtype).max)
    valid_primes = primes[primes > n_colors_prev]
    if len(valid_primes) < min_random_choices:
        valid_primes = primes[-min_random_choices:]
    n_colors = rng.choice(valid_primes)

    extended_colors = cached_colors[:ncolors]
    
    indices = rng.shuffle(np.arange(len(extended_colors)))
    
    new_colormap = LabelColormap(
        name=colormap.name,
        colors=extended_colors[indices],
        controls=np.linspace(0, 1, len(extended_colors) + 1),
        interpolation='zero',
        background_value=colormap.background_value,
    )
    return new_colormap
```

As `n_colors` may be twice as big as `len(colormap.colors)` I suggest this solution that uses existing cache.

Reply:
@Czaki this feels like a premature optimization. In 99% of use cases this will be a uint8 array and it takes 90µs to run the whole function. For uint16 it's 4ms — still acceptable when pressing the shuffle button, I think. Using the cache relies on an implementation detail. I had to do a bit of fixing on the code — `ncolors` typo on line 541 (should be `n_colors`), and `shuffle` works in-place and returns None, so the indices need to be created first, then shuffled in a separate line. After doing that, this version has a very small performance gain (92->82µs), and even smaller for uint16 (4.44->4.29ms). So I think it is better to not rely on the private cache, which is more likely to change in the future and thus be more difficult to maintain:

```
In [3]: from napari.layers import Labels

In [4]: layer = Labels(np.random.randint(50, size=(10, 10)))

In [6]: from napari.utils.colormaps import colormap_utils as cu

In [7]: %timeit cu.shuffle_and_extend_colormap(layer.colormap, 42)
92.1 µs ± 1.07 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [8]: layer.num_colors = 60000

In [9]: %timeit cu.shuffle_and_extend_colormap(layer.colormap, 42)
4.44 ms ± 19.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [11]: layer = Labels(np.random.randint(50, size=(10, 10)))

In [19]: %timeit cu.shuffle_and_extend_colormap2(layer.colormap, 42)
82.4 µs ± 78 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [20]: layer.num_colors = 60000

In [21]: %timeit cu.shuffle_and_extend_colormap2(layer.colormap, 42)
4.29 ms ± 1.77 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```